Src/BootCamp.Chapter/Credentials.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.IO;
+
+namespace BootCamp.Chapter
+{
+    public class Credentials
+    {
+        private const string Separator = ",";
+
+        public string CredentialsFile { get; } = "credentials.db";
+
+        public Credentials()
+        {
+        }
+
+        public Credentials(string credentialsFile)
+        {
+            CredentialsFile = credentialsFile ?? throw new ArgumentNullException(nameof(credentialsFile));
+        }
+
+        private bool FindUser(User user)
+        {
+            string line;
+            bool found = false;
+
+            StreamReader reader;
+            try
+            {
+                reader = new StreamReader(CredentialsFile);
+                while ((line = reader.ReadLine()) != null)
+                {
+                    var isValid = TryParse(line, out User credentials);
+                    if (isValid && user.Equals(credentials))
+                    {
+                        found = true;
+                        break;
+                    }
+                }
+            }
+            catch (Exception)
+            {
+                throw new InvalidCredentialsDbFile($"There was an error while trying to work with {CredentialsFile}");
+            }
+            reader?.Close();
+            return found;
+        }
+
+        private bool AddUser(User user)
+        {
+            if (FindUser(user))
+            {
+                throw new UserAllreadyExistsException("User already exists!");
+            }
+            StreamWriter writer;
+            try
+            {
+                writer = new StreamWriter(CredentialsFile, true);
+                writer.WriteLine($"{user.Name},{user.Password}");
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidCredentialsDbFile($"There was an error while trying to work with {CredentialsFile}", ex);
+            }
+
+            writer?.Close();
+            return true;
+        }
+
+        private static bool TryParse(string input, out User user)
+        {
+            user = default;
+
+            if (!StringOps.IsValid(input))
+            {
+                return false;
+            }
+
+            var fields = input.Split(Separator);
+            const int fieldsNumber = 2;
+
+            var isValid = fields.Length == fieldsNumber || StringOps.IsValid(fields[0]) || StringOps.IsValid(fields[1]);
```

Comment:
wouldn't these be `&&`s also?

Reply:
This one led to refactoring the whole class.
I fixed in [2a4af80](https://github.com/csinn/CSharp-From-Zero-To-Hero/pull/232/commits/2a4af80f11d86f38c8efd0a1debe8b0fd5ae911e)